### PR TITLE
nixVersions.nix_2_10: init at 2.10.0

### PR DESCRIPTION
**Pull request by NickCao**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
kept as a draft until https://github.com/NixOS/nix/issues/6789 is resolved

**Comment by NickCao** [comment]
Result of `nixpkgs-review pr 181180` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>rnix-lsp</li>
  </ul>
</details>
<details>
  <summary>62 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cached-nix-shell</li>
    <li>cachix</li>
    <li>colmena</li>
    <li>comma</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>dep2nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>go2nix</li>
    <li>hci</li>
    <li>hercules-ci-agent</li>
    <li>hydra_unstable</li>
    <li>libnixxml</li>
    <li>lispPackages.quicklisp-to-nix</li>
    <li>lispPackages.quicklisp-to-nix-system-info</li>
    <li>nix</li>
    <li>nix-binary-cache</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-doc</li>
    <li>nix-du</li>
    <li>nix-eval-jobs</li>
    <li>nix-index</li>
    <li>nix-pin</li>
    <li>nix-plugins</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-serve</li>
    <li>nix-simple-deploy</li>
    <li>nix-template</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nixStatic</li>
    <li>nixVersions.nix_2_9</li>
    <li>nixos-generators</li>
    <li>nixos-rebuild</li>
    <li>nixos-shell</li>
    <li>nixpkgs-review</li>
    <li>node2nix</li>
    <li>npins</li>
    <li>nuget-to-nix</li>
    <li>nvfetcher</li>
    <li>prefetch-yarn-deps</li>
    <li>python310Packages.nix-kernel</li>
    <li>python39Packages.nix-kernel</li>
    <li>terranix</li>
    <li>update-nix-fetchgit</li>
    <li>vgo2nix</li>
    <li>vulnix</li>
    <li>wp4nix</li>
    <li>yarn2nix</li>
  </ul>
</details>

**Comment by roberth** [comment]
Please split into three commits when the blocker is resolved.

1. `nixVersions.nix_2_9: 2.9.1 -> 2.9.2`
2. `nixVersions.nix_2_10: init`
3. `nix: 2.9 -> 2.10`

I don't know if we actually have a convention for the last one.
cc @lovesegfault

**Comment by lovesegfault** [comment]
> Please split into three commits when the blocker is resolved.
> 
>     1. `nixVersions.nix_2_9: 2.9.1 -> 2.9.2`
> 
>     2. `nixVersions.nix_2_10: init at <VERSION>`
> 
>     3. `nix: 2.9 -> 2.10`
> 
> 
> I don't know if we actually have a convention for the last one. cc @lovesegfault

Yup, I agree that's how it should be done.

This is also missing an update to fallback paths.

**Comment by NickCao** [comment]
> This is also missing an update to fallback paths.

That one is from `maintainers/upload-release.pl`, normally @edolstra would fix that after version bumps.

**Comment by NickCao** [comment]
In case anyone else also need a standalone script to update `nix-fallback-paths.nix`
```perl
#! /usr/bin/env nix-shell
#! nix-shell -i perl -p perl perlPackages.LWPUserAgent perlPackages.LWPProtocolHttps

use strict;
use JSON::PP;
use LWP::UserAgent;

my $evalId = $ARGV[0] or die "Usage: $0 EVAL-ID\n";

sub fetch {
    my ($url, $type) = @_;

    my $ua = LWP::UserAgent->new;
    $ua->default_header('Accept', $type) if defined $type;

    my $response = $ua->get($url);
    die "could not download $url: ", $response->status_line, "\n" unless $response->is_success;

    return $response->decoded_content;
}

my $evalUrl = "https://hydra.nixos.org/eval/$evalId";

sub getStorePath {
    my ($jobName) = @_;
    my $buildInfo = decode_json(fetch("$evalUrl/job/$jobName", 'application/json'));
    return $buildInfo->{buildoutputs}->{out}->{path} or die "cannot get store path for '$jobName'";
}

print("{\n" .
      "  x86_64-linux = \"" . getStorePath("build.x86_64-linux") . "\";\n" .
      "  i686-linux = \"" . getStorePath("build.i686-linux") . "\";\n" .
      "  aarch64-linux = \"" . getStorePath("build.aarch64-linux") . "\";\n" .
      "  x86_64-darwin = \"" . getStorePath("build.x86_64-darwin") . "\";\n" .
      "  aarch64-darwin = \"" . getStorePath("build.aarch64-darwin") . "\";\n" .
      "}\n");
```

**Comment by NickCao** [comment]
Result of `nixpkgs-review pr 181180` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>74 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cached-nix-shell</li>
    <li>cachix</li>
    <li>colmena</li>
    <li>comma</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>dep2nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>go2nix</li>
    <li>haskellPackages.cabal2nix-unstable</li>
    <li>haskellPackages.cachix</li>
    <li>haskellPackages.hercules-ci-agent</li>
    <li>haskellPackages.hercules-ci-cli</li>
    <li>haskellPackages.hercules-ci-cnix-expr</li>
    <li>haskellPackages.hercules-ci-cnix-store</li>
    <li>haskellPackages.nix-paths</li>
    <li>haskellPackages.nvfetcher</li>
    <li>haskellPackages.update-nix-fetchgit</li>
    <li>hci</li>
    <li>hercules-ci-agent</li>
    <li>hydra_unstable</li>
    <li>libnixxml</li>
    <li>lispPackages.quicklisp-to-nix</li>
    <li>lispPackages.quicklisp-to-nix-system-info</li>
    <li>nix (nixVersions.nix_2_10 ,nixVersions.stable ,nixVersions.unstable)</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-doc</li>
    <li>nix-du</li>
    <li>nix-eval-jobs</li>
    <li>nix-index</li>
    <li>nix-pin</li>
    <li>nix-plugins</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-serve</li>
    <li>nix-simple-deploy</li>
    <li>nix-template</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nixStatic</li>
    <li>nixVersions.nix_2_9</li>
    <li>nixos-generators</li>
    <li>nixos-rebuild</li>
    <li>nixos-shell</li>
    <li>nixpkgs-review</li>
    <li>node2nix</li>
    <li>npins</li>
    <li>nuget-to-nix</li>
    <li>nvfetcher</li>
    <li>prefetch-yarn-deps</li>
    <li>python310Packages.nix-kernel</li>
    <li>python39Packages.nix-kernel</li>
    <li>rnix-lsp</li>
    <li>terranix</li>
    <li>tests.haskell.cabalSdist.hercules-ci-cnix-store</li>
    <li>tests.testers.nixosTest-example</li>
    <li>tests.trivial-builders.references</li>
    <li>update-nix-fetchgit</li>
    <li>vgo2nix</li>
    <li>vulnix</li>
    <li>wp4nix</li>
    <li>yarn2nix</li>
  </ul>
</details>

And one remaining problem: `tests/gc-non-blocking.sh` is flaky as shown in ofborg.

**Comment by lovesegfault** [comment]
@NickCao I think we have two options here:
1. Wait a few days to see if a new release comes out fixing those bugs, then merge this as-is.
2. Add `nix_2_10`, but don't make it the default, also removing the change to fallback paths. Make it the default and update fallback paths in the future, once the issues are ironed out.

Since you're doing the work, I'd like to know what you prefer. Is there another avenue I'm missing?

**Comment by NickCao** [comment]
Reused the same trick that retries the mysterious test failures on darwin, but I believe this one on linux is more of a bad test than a bad nix (nearly always reproducible on a busy machine).

**Comment by edolstra** [comment]
2.10.2 is out, here are the fallback paths:
```
commit fe30ff4ba425b2e18cb630c6f660668720c31136 (HEAD -> master)
Author: Eelco Dolstra <edolstra@gmail.com>
Date:   Thu Jul 14 17:10:57 2022 +0200

    nix-fallback-paths.nix: Update to 2.10.2

diff --git a/nixos/modules/installer/tools/nix-fallback-paths.nix b/nixos/modules/installer/tools/nix-fallback-paths.nix
index bd09724ebfa..2df7b1ae2d2 100644
--- a/nixos/modules/installer/tools/nix-fallback-paths.nix
+++ b/nixos/modules/installer/tools/nix-fallback-paths.nix
@@ -1,7 +1,7 @@
 {
-  x86_64-linux = "/nix/store/6g4fla3vkcxihph282a0v3cd10709y7c-nix-2.9.1";
-  i686-linux = "/nix/store/j143221z44469zx21f5m9a47x7y1jpr5-nix-2.9.1";
-  aarch64-linux = "/nix/store/c4z3vy1sgm49la8bvmdrrpssgk4iw2nk-nix-2.9.1";
-  x86_64-darwin = "/nix/store/cqdwb7khf6zg94bz7lnvfjqx6z775qaw-nix-2.9.1";
-  aarch64-darwin = "/nix/store/1brkxcs287n1px2i4fq39l7h51hjv0f8-nix-2.9.1";
+  x86_64-linux = "/nix/store/qx384jvdjk2vzd2wpr9pmkcimk12g3as-nix-2.10.2";
+  i686-linux = "/nix/store/8mhknp6flnlnbj11zmskb3972ia1mdl5-nix-2.10.2";
+  aarch64-linux = "/nix/store/digzqw2vgdsg0fdzf7qlwbl4xii9i458-nix-2.10.2";
+  x86_64-darwin = "/nix/store/w62jxmpvjykw707qr9s5nxi90r35xqmd-nix-2.10.2";
+  aarch64-darwin = "/nix/store/6pjrn22f57n854x88a9kkbqnxw1yai73-nix-2.10.2";
 }
```

We should remove Nix 2.9 since we decided that we shouldn't keep old versions around unless there is a compelling reason to do so.

**Comment by NickCao** [comment]
Oh wait, there's a few more packages pinned on nix 2.9. We really should keep that for a little longer.

**Comment by lovesegfault** [comment]
> 2.10.2 is out, here are the fallback paths:

Thanks!
 
> We should remove Nix 2.9 since we decided that we shouldn't keep old versions around unless there is a compelling reason to do so.

We need to wait for the hercules-ci pkgs and cachix to catch up, then we can remove it.

**Comment by Kranzes** [comment]
Can someone create a follow-up PR for enabling LTO support if the stdenv (gcc) supports it? (I am not sure if LTO works on Darwin or 32bit)

File: pkgs/tools/package-management/nix/default.nix [hunk 1]
```diff
@@ -47,59 +47,31 @@ in lib.makeExtensible (self: {
     };
   }).override { boehmgc = boehmgc-nix_2_3; };
 
-  nix_2_4 = common {
-    version = "2.4";
-    sha256 = "sha256-op48CCDgLHK0qV1Batz4Ln5FqBiRjlE6qHTiZgt3b6k=";
-    # https://github.com/NixOS/nix/pull/5537
-    patches = [ ./patches/install-nlohmann_json-headers.patch ];
-  };
+  nix_2_4 = throw "nixVersions.nix_2_4 has been removed";
```
Comment (Mindavi) [inline pkgs/tools/package-management/nix/default.nix:50]: Shouldn't these be in aliases.nix?

**Comment by NickCao** [comment]
@edolstra I noticed that the fallback paths for `x86_64-linux` and `x86_64-darwin` is missing in `cache.nixos.org`, may you fix this?

File: pkgs/tools/package-management/nix/default.nix (hunk 1 above)
Comment (NickCao) [reply pkgs/tools/package-management/nix/default.nix:50]: There would be direct references to `nixVersions.*` out there, leave these throws here makes me feel more comfortable.

**Comment by NickCao** [comment]
Also `[amazonImage.x86_64-linux](https://hydra.nixos.org/job/nixos/unstable-small/nixos.amazonImage.x86_64-linux)` or in general `make-disk-image` is broken by auto-chroot-store.

File: pkgs/tools/package-management/nix/common.nix [hunk 2]
```diff
@@ -176,6 +177,7 @@ self = stdenv.mkDerivation {
   installFlags = [ "sysconfdir=$(out)/etc" ];
 
   doInstallCheck = true;
+  installCheckTarget = if atLeast210 then "installcheck" else null;
```
Comment (SuperSandro2000) [inline pkgs/tools/package-management/nix/common.nix:180]:
```suggestion
  installCheckTarget = lib.optionalString atLeast210 "installcheck";
```
nit

Comment (NickCao) [reply pkgs/tools/package-management/nix/common.nix:180]: I was trying to reduce the number of rebuild, but given that a lot of old versions were removed, this wouldn't have made a deep dent.

**Comment by edolstra** [comment]
Here is 2.10.3:
```
commit d718fde50180e5ab5a4a2c666907b8ea8c714537 (HEAD -> master)
Author: Eelco Dolstra <edolstra@gmail.com>
Date:   Fri Jul 15 18:42:12 2022 +0200

    nix-fallback-paths.nix: Update to 2.10.3

diff --git a/nixos/modules/installer/tools/nix-fallback-paths.nix b/nixos/modules/installer/tools/nix-fallback-paths.nix
index 2df7b1ae2d2..0035ceca6fc 100644
--- a/nixos/modules/installer/tools/nix-fallback-paths.nix
+++ b/nixos/modules/installer/tools/nix-fallback-paths.nix
@@ -1,7 +1,7 @@
 {
-  x86_64-linux = "/nix/store/qx384jvdjk2vzd2wpr9pmkcimk12g3as-nix-2.10.2";
-  i686-linux = "/nix/store/8mhknp6flnlnbj11zmskb3972ia1mdl5-nix-2.10.2";
-  aarch64-linux = "/nix/store/digzqw2vgdsg0fdzf7qlwbl4xii9i458-nix-2.10.2";
-  x86_64-darwin = "/nix/store/w62jxmpvjykw707qr9s5nxi90r35xqmd-nix-2.10.2";
-  aarch64-darwin = "/nix/store/6pjrn22f57n854x88a9kkbqnxw1yai73-nix-2.10.2";
+  x86_64-linux = "/nix/store/3af6g226v4hsv6x7xzh23d6wqyq0nzjp-nix-2.10.3";
+  i686-linux = "/nix/store/43xxh2jip6rpdhylc5z9a5fxx54dw206-nix-2.10.3";
+  aarch64-linux = "/nix/store/6qw3r57nra08ars8j8zyj3fl8lz4cvnd-nix-2.10.3";
+  x86_64-darwin = "/nix/store/3b7qrm0qjw57fmznrsvm0ai568i89hc2-nix-2.10.3";
+  aarch64-darwin = "/nix/store/gp7k17iy1n7hgf97qwnxw28c6v9nhb1i-nix-2.10.3";
 }
```

**Comment by NickCao** [comment]
> @edolstra I noticed that the fallback paths for `x86_64-linux` and `x86_64-darwin` is missing in `cache.nixos.org`, may you fix this?

This appears again on 2.10.3, turns out to be some bad negative caching on our CDN.
`https://cache.nixos.org/3af6g226v4hsv6x7xzh23d6wqyq0nzjp.narinfo`
`https://aseipp-nix-cache.global.ssl.fastly.net/3af6g226v4hsv6x7xzh23d6wqyq0nzjp.narinfo`

**Comment by vcunat** [comment]
Updated to 2.10.3 in PR #181625

**Comment by milahu** [comment]
> We should remove Nix 2.9 since we decided that we shouldn't keep old versions around unless there is a compelling reason to do so.

i need nix 2.7 as workaround for https://github.com/NixOS/nix/issues/6572 (milestone: nix 2.11)

**Comment by vcunat** [comment]
Honestly, I didn't understand why these removals happened.  It seemed cheap enough to keep a few older versions.  Nix is a very important tool here.

**Comment by roberth** [comment]
> Nix is a very important tool here.

There's a conflict of interest here. Nixpkgs contributors need multiple versions, whereas Nix developers only want to bother with as few versions as possible. Although we should listen to upstream but we should act in the interest of Nixpkgs users and maintainers.

> > unless there is a compelling reason to do so.

> i need nix 2.7 as workaround for NixOS/nix#6572 (milestone: nix 2.11)

Voila.

I support the reversal of the removals.

**Comment by lovesegfault** [comment]
I'm okay with reversing them, but I can't stand this back-and-forth anymore. Every time a new Nix release happens we go over the same debate on whether to keep the old versions or not.

I don't like living in Groundhog day.

**Comment by milahu** [comment]
@NickCao can you please revert b2eea05b06baa2886039d61485d37c338ad3d578

**Comment by NickCao** [comment]
> @NickCao can you please revert [b2eea05](https://github.com/NixOS/nixpkgs/commit/b2eea05b06baa2886039d61485d37c338ad3d578)

May I ask what requires an old version of nix?

**Comment by milahu** [comment]
see above https://github.com/NixOS/nixpkgs/pull/181180#issuecomment-1205459646

**Comment by NickCao** [comment]
I haven't seen that particular issue for a long time, I thought we've get rid of it.

**Comment by amjoseph-nixpkgs** [comment]
I am experiencing #181180 right now and it is causing me major headaches.

The removal of old versions of nix makes me less willing to "live at HEAD" where I can notice regressions as soon as they happen.

**Comment by SuperSandro2000** [comment]
> I am experiencing #181180 right now and it is causing me major headaches.

You mean https://github.com/NixOS/nixpkgs/pull/181180#issuecomment-1205459646 ?

**Comment by amjoseph-nixpkgs** [comment]
> You mean [#181180 (comment)](https://github.com/NixOS/nixpkgs/pull/181180#issuecomment-1205459646) ?

Sorry, I meant https://github.com/NixOS/nix/issues/6572

> Also 2.7 was already added back in #186798

Thank you for merging that just eight hours after my comment!

**Comment by Profpatsch** [comment]
https://github.com/NixOS/nix/issues/6572 is a critical bug that hasn’t been addressed in 6 months, and there is no way to pin nix to a working version from nixpkgs.

**Comment by Profpatsch** [comment]
I revived nix 2.6 https://github.com/NixOS/nixpkgs/pull/200357

**Pull request merged**